File: packages/guards/src/tool-call-sequence.ts
import type { Guard, GuardResult, GuardContext } from 'open-guardrail-core';

interface ToolCallSequenceOptions {
  action: 'block' | 'warn';
  /** Pairs of [first, second] that are dangerous in sequence */
  dangerousSequences?: string[][];
}

const DEFAULT_DANGEROUS_SEQUENCES: string[][] = [
  ['read', 'delete'],
  ['query', 'drop'],
  ['list', 'delete'],
  ['get', 'send'],
  ['fetch', 'upload'],
  ['read', 'execute'],
  ['download', 'execute'],
  ['search', 'modify'],
  ['access', 'transfer'],
  ['view', 'export'],
];

function extractActions(text: string): string[] {
  const actions: string[] = [];
  const patterns = [
    /(?:tool_call|function_call|use_tool|Action)[\s:]*(\w+)/gi,
    /<tool>\s*(\w+)/gi,
    /(\w+)\s*\(/g,
  ];
  for (const pattern of patterns) {
    let match: RegExpExecArray | null;
    while ((match = pattern.exec(text)) !== null) {
      actions.push(match[1].toLowerCase());
    }
  }
  return actions;
}

export function toolCallSequence(options: ToolCallSequenceOptions): Guard {
  const sequences = options.dangerousSequences ?? DEFAULT_DANGEROUS_SEQUENCES;

  return {
    name: 'tool-call-sequence',
    version: '0.1.0',
    description: 'Detects dangerous sequential tool call patterns',
    category: 'security',
    supportedStages: ['input', 'output'],
    async check(text: string, _ctx: GuardContext): Promise<GuardResult> {
      const start = performance.now();
      const actions = extractActions(text);
      const matched: string[][] = [];

      for (const [first, second] of sequences) {
        const firstIdx = actions.indexOf(first);
        if (firstIdx === -1) continue;
        for (let i = firstIdx + 1; i < actions.length; i++) {
          if (actions[i] === second) {
            matched.push([first, second]);
            break;
          }
        }
      }

      const triggered = matched.length > 0;
      return {
        guardName: 'tool-call-sequence',
        passed: !triggered,
        action: triggered ? options.action : 'allow',
        latencyMs: Math.round(performance.now() - start),
        details: triggered
          ? { matchedSequences: matched, extractedActions: actions }
          : undefined,
      };
    },
  };
}

File: python/open_guardrail/guards/tool_call_sequence.py
"""Detects dangerous sequential tool call patterns."""
from __future__ import annotations

import re
import time
from typing import List, Optional

from open_guardrail.core import GuardResult

_DEFAULT_DANGEROUS_SEQUENCES: List[List[str]] = [
    ["read", "delete"],
    ["query", "drop"],
    ["list", "delete"],
    ["get", "send"],
    ["fetch", "upload"],
    ["read", "execute"],
    ["download", "execute"],
    ["search", "modify"],
    ["access", "transfer"],
    ["view", "export"],
]

_ACTION_PATTERNS = [
    re.compile(r"(?:tool_call|function_call|use_tool|Action)[\s:]*(\w+)", re.IGNORECASE),
    re.compile(r"<tool>\s*(\w+)", re.IGNORECASE),
    re.compile(r"(\w+)\s*\("),
]


def _extract_actions(text: str) -> List[str]:
    actions: List[str] = []
    for pattern in _ACTION_PATTERNS:
        for m in pattern.finditer(text):
            actions.append(m.group(1).lower())
    return actions


class _ToolCallSequence:
    def __init__(
        self,
        *,
        action: str = "block",
        dangerous_sequences: Optional[List[List[str]]] = None,
    ) -> None:
        self.name = "tool-call-sequence"
        self.action = action
        self.sequences = dangerous_sequences or _DEFAULT_DANGEROUS_SEQUENCES

    def check(self, text: str, stage: str = "input") -> GuardResult:
        start = time.perf_counter()
        actions = _extract_actions(text)
        matched: List[List[str]] = []

        for first, second in self.sequences:
            try:
                first_idx = actions.index(first)
            except ValueError:
                continue
            for i in range(first_idx + 1, len(actions)):
                if actions[i] == second:
                    matched.append([first, second])
                    break

        triggered = len(matched) > 0
        elapsed = (time.perf_counter() - start) * 1000
        return GuardResult(
            guard_name="tool-call-sequence",
            passed=not triggered,
            action=self.action if triggered else "allow",
            message=f"Dangerous tool sequence detected: {matched}" if triggered else None,
            latency_ms=round(elapsed, 2),
            details={"matched_sequences": matched, "extracted_actions": actions} if triggered else None,
        )


def tool_call_sequence(
    *,
    action: str = "block",
    dangerous_sequences: Optional[List[List[str]]] = None,
) -> _ToolCallSequence:
    return _ToolCallSequence(
        action=action,
        dangerous_sequences=dangerous_sequences,
    )
